streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java

```diff
@@ -0,0 +1,390 @@
+            tasksAndActions.clear();
+            return tasksAndActionsToProcess;
+        }
+
+        private void addTask(final Task task) {
+            catchException(() -> {
+                final Task.State state = task.state();
+                switch (state) {
+
+                    case CREATED:
+                        task.initializeIfNeeded();
+                        // Todo: catch exceptions and clear task timeout
+                        if (isStateless(task)) {
+                            addTaskToRestoredTasks((StreamTask) task);
+                        } else {
+                            updatingTasks.put(task.id(), task);
+                        }
+                        break;
+
+                    case SUSPENDED:
+                        task.resume();
+                        break;
+
+                    default:
+                        throw new IllegalStateException("Illegal state " + state + " while adding to the state updater. "
+                            + BUG_ERROR_MESSAGE);
+                }
+                return null;
+            }, task);
+        }
+
+        private boolean isStateless(final Task task) {
+            return task.changelogPartitions().isEmpty() && task.isActive();
+        }
+
+        private void removeTask(final Task task) {
+            catchException(() -> {
+                final Collection<TopicPartition> changelogPartitions = task.changelogPartitions();
+                if (!changelogPartitions.isEmpty()) {
+                    updatingTasks.remove(task.id());
+                    task.stateManager().checkpoint();
+                    changelogReader.unregister(changelogPartitions);
+                }
+                if (task.isActive()) {
+                    removeTaskFromRestoredTasks((StreamTask) task);
+                }
+                task.suspend();
+                task.closeClean();
+                return null;
+            }, task);
+        }
+
+        private void catchException(final Supplier<Void> codeToCheck, final Task task) {
+            try {
+                codeToCheck.get();
+            } catch (final RuntimeException exception) {
+                exceptions.add(exception);
+                updatingTasks.remove(task.id());
+            }
+        }
+
+        private void handleTaskCorruptedException(final TaskCorruptedException taskCorruptedException) {
+            exceptions.add(taskCorruptedException);
+            final Set<TaskId> corruptedTaskIds = taskCorruptedException.corruptedTasks();
+            for (final TaskId taskId : corruptedTaskIds) {
+                updatingTasks.remove(taskId);
+            }
+        }
+
+        private void handleStreamsException(final StreamsException streamsException) {
+            if (streamsException.taskId().isPresent()) {
+                final Task task = updatingTasks.get(streamsException.taskId().get());
+                task.stateManager().checkpoint();
```

Comment:
I thought the idea is to checkpoint if we need to interrupt restoration because we actually updated the state. Otherwise we would reread the changelog topic from the position before we started restoration. But now that I think about it with eos that might be a problem, right? Probably, I am missing something here.